controllers/argocd/notifications/util.go

```diff
@@ -0,0 +1,664 @@
+
+	// configure default notifications triggers
+
+	notificationsConfig["trigger.on-created"] = `- description: Application is created.
+  oncePer: app.metadata.name
+  send:
+  - app-created
+  when: "true"`
+
+	notificationsConfig["trigger.on-deleted"] = `- description: Application is deleted.
+  oncePer: app.metadata.name
+  send:
+  - app-deleted
+  when: app.metadata.deletionTimestamp != nil`
+
+	notificationsConfig["trigger.on-deployed"] = `- description: Application is synced and healthy. Triggered once per commit.
+  oncePer: app.status.operationState.syncResult.revision
+  send:
+  - app-deployed
+  when: app.status.operationState.phase in ['Succeeded'] and app.status.health.status
+      == 'Healthy'`
+
+	notificationsConfig["trigger.on-health-degraded"] = `- description: Application has degraded
+  send:
+  - app-health-degraded
+  when: app.status.health.status == 'Degraded'`
+
+	notificationsConfig["trigger.on-sync-failed"] = `- description: Application syncing has failed
+  send:
+  - app-sync-failed
+  when: app.status.operationState.phase in ['Error', 'Failed']`
+
+	notificationsConfig["trigger.on-sync-running"] = `- description: Application is being synced
+  send:
+  - app-sync-running
+  when: app.status.operationState.phase in ['Running']`
+
+	notificationsConfig["trigger.on-sync-status-unknown"] = `- description: Application status is 'Unknown'
+  send:
+  - app-sync-status-unknown
+  when: app.status.sync.status == 'Unknown'`
+
+	notificationsConfig["trigger.on-sync-succeeded"] = `- description: Application syncing has succeeded
+  send:
+  - app-sync-succeeded
+  when: app.status.operationState.phase in ['Succeeded']`
+
+	return notificationsConfig
+}
+
+func GetNotificationsCommand(cr *argoprojv1a1.ArgoCD) []string {
+
+	cmd := make([]string, 0)
+	cmd = append(cmd, "argocd-notifications")
+
+	cmd = append(cmd, "--loglevel")
+	cmd = append(cmd, argoutil.GetLogLevel(cr.Spec.Notifications.LogLevel))
+
+	return cmd
+}
+
+// GetNotificationsResources will return the ResourceRequirements for the Notifications container.
+func GetNotificationsResources(cr *argoprojv1a1.ArgoCD) corev1.ResourceRequirements {
+	resources := corev1.ResourceRequirements{}
+
+	// Allow override of resource requirements from CR
+	if cr.Spec.Notifications.Resources != nil {
+		resources = *cr.Spec.Notifications.Resources
+	}
+
+	return resources
+}
+
+func GetArgoCDNotificationsControllerReplicas(nr *NotificationsReconciler) *int32 {
+	if nr.Instance.Spec.Notifications.Replicas != nil && *nr.Instance.Spec.Notifications.Replicas >= 0 {
+		return nr.Instance.Spec.Notifications.Replicas
+	}
+
+	return nil
+}
+
+func UpdateIfChanged(existingVal, desiredVal interface{}, extraAction func(), changed *bool) {
+	if !reflect.DeepEqual(existingVal, desiredVal) {
+		reflect.ValueOf(existingVal).Elem().Set(reflect.ValueOf(desiredVal).Elem())
+		if extraAction != nil {
+			extraAction()
+		}
+		*changed = true
+	}
+}
```

Comment:
This function belongs in a common package accessible to all controller packages

I will move it in a separate PR